lib/color_spinor_project_dd.cu

```diff
@@ -0,0 +1,143 @@
+#include <tuple>
+#include <memory>
+#include <color_spinor_field.h>
+#include <color_spinor_field_order.h>
+#include <index_helper.cuh>
+#include <blas_quda.h>
+#include <instantiate.h>
+#include <domain_decomposition_helper.cuh>
+#include <tunable_nd.h>
+#include <kernels/color_spinor_project_dd.cuh>
+
+namespace quda
+{
+
+  template <typename Float, typename DDArg, int nSpin, int nColor, typename Order>
+  class ProjectDD : public TunableKernel2D
+  {
+    using Arg = ProjectDDArg<Float, DDArg, nSpin, nColor, Order>;
+    ColorSpinorField &out;
+
+    bool advanceSharedBytes(TuneParam &) const { return false; } // Don't tune shared mem
+    unsigned int minThreads() const { return out.VolumeCB(); }
+
+  public:
+    ProjectDD(ColorSpinorField &out) : TunableKernel2D(out, out.SiteSubset()), out(out)
+    {
+      strcat(aux, out.AuxString().c_str());
+      switch (out.DD().type) {
+      case QUDA_DD_NO: strcat(aux, ",DDNo"); break;
+      case QUDA_DD_RED_BLACK: strcat(aux, ",DDRedBlack"); break;
+      default: errorQuda("DD type %d not implemented", out.DD().type);
+      }
+
+      apply(device::get_default_stream());
+    }
+
+    void apply(const qudaStream_t &stream)
+    {
+      constexpr bool enable_host = true;
+      TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+      launch<ProjectDD_, enable_host>(tp, stream, Arg(out));
+    }
+
+    long long bytes() const { return out.Bytes(); }
+  };
+
+  template <typename P, typename DDArg> void projectDD(P &p, DDArg &dd, const ColorSpinorField &meta)
+  {
+    Coord<4> coord;
+    int X[4] = {meta.full_dim(0), meta.full_dim(1), meta.full_dim(2), meta.full_dim(3)};
+    int commCoord[4] = {comm_coord(0) * X[0], comm_coord(1) * X[1], comm_coord(2) * X[2], comm_coord(3) * X[3]};
+
+    for (int parity = 0; parity < p.Nparity(); parity++) {
+      for (int x_cb = 0; x_cb < p.VolumeCB(); x_cb++) {
+        getCoords(coord, x_cb, X, parity);
+        for (int i = 0; i < coord.size(); i++) { coord.gx[i] = commCoord[i] + coord.x[i]; }
+
+        if (dd.isZero(coord)) {
+          for (int s = 0; s < p.Nspin(); s++)
+            for (int c = 0; c < p.Ncolor(); c++) p(parity, x_cb, s, c) = 0;
+        }
+      }
+    }
+  }
+
+  template <typename Float, typename DDArg, int nSpin, int nColor, typename Order>
+  void genericProjectDD(ColorSpinorField &a)
+  {
+    /* Reference CPU implementation
```

Comment:
Remove this, unless there's a reason to uncomment it

Reply:
I removed it